### ISSUE TITLE
Cannot open include file: 'fmpz-conversions.h': No such file or directory

### ISSUE THREAD

**Issue by kaba2** [BrianGladman/flint#2]
When trying to build the solution in Visual Studio 2022, I get:

"Cannot open include file: 'fmpz-conversions.h': No such file or directory"

I can only find `fmpz-conversions-gc.in`, `fmpz-conversions-reentrant.in`, and `fmpz-conversions-single.in` in Flint root directory, and no `fmpz-conversions.h`.

Is there something I should do before building the solution file?

**Comment by BrianGladman** [comment]
Hi, I have just added a Visual Studio 20l22 build.   Could you see if this works for you.

**Comment by kaba2** [comment]
The same problem occurs. I think `.in` files are by convention files that CMake scripts convert to another file by substituting CMake variables. Perhaps these are a new addition?

**Comment by kaba2** [comment]
The file `fmpz-conversions.h` is produced on line 185 in `CMakeLists.txt`:

```cmake
configure_file(
    fmpz-conversions-${MEMORY_MANAGER}.in
    fmpz-conversions.h
    COPYONLY
)
```

So it seems that `cmake` would somehow be have to run before to produce the file.

**Comment by BrianGladman** [comment]
My build makes no use whatsoever of cmake.  Just load flint.sln and build either or both the DLL and the static library.

**Comment by kaba2** [comment]
Here is what I am doing:
* I download [latest trunk](https://github.com/BrianGladman/flint).
* Open `flint.sln`.
* Select `Release` and `x64`.
* Right click on `dll_flint` project -> `Project only `-> `Build only dll_flint`.

Error: Cannot open include file: 'fmpz-conversions.h': No such file or directory. Indeed, there is no file `fmpz-conversions.h` in this repository. If you don't get this error, could you be having such a file somewhere else on your include path?

**Comment by BrianGladman** [comment]
OK, that locates the issue. I have now added th is file that was missing. Hopefully that will fix it for you.

**Comment by kaba2** [comment]
Thanks. The project builds almost to the end. But then there is a new error: "Cannot open include file: 'fft_tuning.h': No such file or directory". That file is similar to the previous one. In `CMakeLists.txt`, it is generated on lines 197-207:

```cmake
if(CMAKE_SIZEOF_VOID_P EQUAL 8)
    configure_file(
        fft_tuning64.in
        fft_tuning.h
        COPYONLY
    )   
elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
    configure_file(
        fft_tuning32.in
        fft_tuning.h
        COPYONLY
    )
endif()
```

In this case the correct file seems to depend on whether you want to build 32-bit or 64-bit.

**Comment by kaba2** [comment]
I can fix the previous problem by copying `fft_tuning64.in` to `fft_tuning.h`, and therefore committing to a 64-bit build. 

After that, there is one more error:  there is no file `fmpz/fmpz.c`. The file `fmpz.c` is produced in `CMakeLists.txt` on lines 191-195:

```cmake
configure_file(
    fmpz/link/fmpz_${MEMORY_MANAGER}.c
    fmpz/fmpz.c
    COPYONLY
)
```

If I copy `fmpz/link/fmpz_gc.g` to `fmpz/fmpz.c`, then the build goes through. Presumably, I could copy any of the three `fmpz_gc.c`, `fmpz_reentrant.c`, or `fmpz_single.c`, but I'm not sure how to choose.

**Comment by BrianGladman** [comment]
I have added some missing files that should solve some of these issues but I can't explain all of them at the moment.

The instructions for building flint are in the file flint\build.vc\readme.txt    Before you run flint\build.vs22\flint.sln it is necessary to load flint\build.vc\flint.sln and run the Python file flint_config.py.  This generates the configuration files that are missing in your build.   I have temporarily added the  missing files to the repository but this is not ideal since they do need to be updated when the files from which they are generated are changed .

**Comment by kaba2** [comment]
I don't think a single static set of files can capture all of the different [builds](https://flintlib.org/doc/building.html). There seems to be a choice of bitness (32/64), and a "memory manager" (3 choices), for a total of 6 different file-sets. When `CMake` is run, it can be provided parameters which fix these choices. CMake then copies the appropriate files over as we have done here manually. This repository defaults to a single choice of these 6 possible choices (perhaps 64-bit gc, depending on what you chose). If I understood correctly, Flint can also be built using `./configure` (followed by make), which accepts similar parameters and does the same copying thing as CMake.

**Comment by BrianGladman** [comment]
See my edit above.  The full instructions for building FLINT are set out in the file: flint\build.vc\readme.txt (I realise that I need to make this file more prominent within this repository).   For this repository flint_config.py replaces cmake and ./configure.

**Comment by kaba2** [comment]
I'll try again with these instructions. 

I hit the first error when opening `flint/build.vc/flint.sln´: "C:\code\testarb2\flint\build.vc\test_one\test_one.vcxproj : error  : The project file could not be loaded. Could not find a part of the path 'C:\code\testarb2\flint\build.vc\test_one\test_one.vcxproj'.  C:\code\testarb2\flint\build.vc\test_one\test_one.vcxproj".

**Comment by kaba2** [comment]
I was able to build flint by first running `python fling_config.py 22` (after removing the extraneous `test_one` project), and building the resulting `flint/build.vs22/flint.sln`. I noticed the following:

* the script chooses `fft_tuning.h` to be `fft_tuning64.h`. 
* the `fmpz-conversions-?.in` variants are all the same file, so for them the choice of `fpmz-conversions.h` does not matter.  
* for `fpmz.c` the script chooses `fmpz/link/fmpz_reentrant.c`. 

Hence, the script makes one choice. I went to look at the script `flint_config.py` and noticed that the parameters could be changed by commenting things out/in on line 53:

```python
# add user choice (duplicate this in _msvc_project.py)
flib_type = 'reentrant' # ('gc', 'reentrant', 'single')
```

Is that the intention? All of these things would preferably need some mentions in the readme. I did not see a parameter for the fft_tuning, and wonder if still building in 32-bit would produce an invalid binary. Or perhaps I should comment out/in lines 253-254?

```python
 # fn = join(flint_dir, 'fft_tuning32.in')
  fn = join(flint_dir, 'fft_tuning64.in')
```

**Comment by BrianGladman** [comment]
Thanks for reporting the issues you found as this will help me in improving the build and its documentation   Comments on what you found follow.

The load failure for test_one can be ignored as it is not a part of the build ((I need it locally for build testing).

I no longer support 32-bit builds because they are obsolete on Windows.  But I left the build code in place in case someone wants to resurrect it.

On the choice of the memory management, I decided to use re-entrant as the default since I have found it the most effective choice on Windows.  But, as you found, the choice can be edited (although it requires a lot of care as changing this has to be reflected on the whole dependency /use build chain (mpir, mpfr, flint, arb, yasm, msieve, gmp-ecm, ...).
